cfgrammar/src/lib/header.rs

```diff
@@ -345,6 +347,132 @@ impl<'input> GrmtoolsSectionParser<'input> {
+}
+
+impl Header {
+    pub fn new() -> Self {
+        Header {
+            contents: MarkMap::new(),
+        }
+    }
+
+    pub fn merge_from(
+        &mut self,
+        other: Self,
+    ) -> Result<(), MergeError<String, Box<(Location, Value)>>> {
+        self.contents.merge_from(other.contents)
+    }
+
+    pub fn contents(&self) -> &MarkMap<String, (Location, Value)> {
+        &self.contents
+    }
+
+    pub fn contents_mut(&mut self) -> &mut MarkMap<String, (Location, Value)> {
+        &mut self.contents
+    }
+
+    /// A wrapper around `HashMap::get`, which checks the `query_mask`
+    /// against the `value.query_bits()`
+    ///
+    /// Regardless of whether this query mask check succeeds or fails,
+    /// if the `contents` contains the key, will mark the key as being used.
+    pub fn query(
+        &mut self,
+        key: &str,
+        query_mask: u16,
+    ) -> Option<Result<(&Location, &Value), HeaderError>> {
+        self.contents_mut().mark_used(&key.to_owned());
+        if let Some((key_loc, value)) = self.contents.get(key) {
+            if value.matches_query_mask(query_mask) {
+                Some(Ok((key_loc, value)))
+            } else {
+                Some(match value {
+                    Value::Flag(_) => Err(HeaderError {
+                        kind: HeaderErrorKind::EntryQueryTypeMismatch,
+                        locations: vec![key_loc.clone()],
+                    }),
+                    Value::Setting(Setting::Num(_, num_loc)) => Err(HeaderError {
+                        kind: HeaderErrorKind::EntryQueryTypeMismatch,
+                        locations: vec![num_loc.clone()],
+                    }),
+                    Value::Setting(Setting::Unitary(Namespaced {
+                        namespace: _,
+                        member: (_, member_loc),
+                    })) => Err(HeaderError {
+                        kind: HeaderErrorKind::EntryQueryTypeMismatch,
+                        locations: vec![member_loc.clone()],
+                    }),
+                    Value::Setting(Setting::Constructor {
+                        ctor:
+                            Namespaced {
+                                namespace: _,
+                                member: (_, _),
+                            },
+                        arg:
+                            Namespaced {
+                                namespace: _,
+                                member: (_, arg_span),
+                            },
+                    }) => Err(HeaderError {
+                        kind: HeaderErrorKind::EntryQueryTypeMismatch,
+                        locations: vec![arg_span.clone()],
+                    }),
+                })
+            }
+        } else {
+            None
+        }
+    }
+}
+
+/// Repr vaues bit representation should not overlap `SettingQuery`,
+/// The first 8 bits is reserved for `ValueQuery`.
+#[repr(u16)]
+pub enum ValueQuery {
```

Comment:
I wonder if we need to expose the maskiness of this to users? The `enum` seems fine (users never need to know what the integer discriminant values are) but when we get `matches_query_mask(u16)` in the public API I do worry a bit about the cognitive load.

A dumb question is: do users ever need to query more than one thing at a time?

Reply:
More than one thing like more than one `key`, or more than one `mask`?

It isn't set up to query more than one key at a time (I never needed it, I don't think we ever do?).
It is used to query more than one value mask value for `yacckind` which can be a unitary `YaccKind::Grmtools` or a constructor `YaccKind::Original(_)`, this is used in the last patch [here](https://github.com/softdevteam/grmtools/pull/549/commits/34272ccaed6e9753612410223db3be8dfa51e7f0#diff-3f17b8629b7e418df92ecbab5e58f5a28a5adcf69d532f5fdb004526d0529ba4R460-R464) if that link works.

```
        if let Some(result) = self.header.query(
            "yacckind",
            SettingQuery::Unitary as u16 | SettingQuery::Constructor as u16,
        )
```

Reply:
Ah, then I really have misunderstood this!

If we do go down this route (and I can be convinced -- I just haven't understood it yet!) I think we would want to find a means that avoids `as`. I have become allergic to `as` because it has surprising effects on some, but not other, types.

Reply:
I don't know of any way to avoid `as` (I will look, maybe i've missed a way being added, but last time I looked it was the only way to convert from an `enum` to it's `#[repr(...)]`) 

At the very least we can wrap it in a function like `fn repr(self) -> u16`  though.